pyls/python_ls.py

```diff
@@ -66,15 +133,34 @@ def capabilities(self):
             'textDocumentSync': lsp.TextDocumentSyncKind.INCREMENTAL,
             'experimental': merge(self._hook('pyls_experimental_capabilities'))
         }
-        log.info("Server capabilities: %s", server_capabilities)
+        log.info('Server capabilities: %s', server_capabilities)
         return server_capabilities
 
-    def initialize(self, root_uri, init_opts, _process_id):
-        self.workspace = Workspace(root_uri, lang_server=self)
-        self.config = config.Config(root_uri, init_opts)
+    def m_initialize(self, processId=None, rootUri=None, rootPath=None, initializationOptions=None, **_kwargs):
+        log.debug('Language server initialized with %s %s %s %s', processId, rootUri, rootPath, initializationOptions)
+        if rootUri is None:
+            rootUri = uris.from_fs_path(rootPath) if rootPath is not None else ''
+
+        self.workspace = Workspace(rootUri, rpc_manager=self.rpc_manager)
+        self.config = config.Config(rootUri, initializationOptions or {})
         self._dispatchers = self._hook('pyls_dispatchers')
         self._hook('pyls_initialize')
 
+        # Get our capabilities
+        return {'capabilities': self.capabilities()}
+
+    def m__cancel_request(self, **kwargs):
```

Comment:
Feels weird to interleave this dispatched methods with the non-dispatched ones (e.g. code_actions). Can you group them all together? Either above or below the others.